src/core/backend_context.cc

```diff
@@ -291,24 +295,56 @@ BackendResponder::Finalize()
                 status),
             "error sending TensorFlow response");
       } else {
-        bool cuda_used = false;
-        status = CopyBuffer(
-            response_output->Name(), pinned_memory_type, pinned_memory_id,
-            response_memory_type, response_memory_type_id, response_byte_size,
-            pinned_buffer + offset, const_cast<void*>(response_buffer), stream_,
-            &cuda_used);
-        need_sync_ |= cuda_used;
+        // Only use parallel CopyBuffer for CPU-CPU copy when worker count > 1
+        if ((AsyncWorkQueue::GetWorkerCount() > 1) &&
+            (response_memory_type == TRITONSERVER_MEMORY_CPU) &&
+            (pinned_memory_type == TRITONSERVER_MEMORY_CPU)) {
+          AsyncWorkQueue::AddTask(std::bind(
+              CopyBufferHandler, response_output->Name(), pinned_memory_type,
+              pinned_memory_id, response_memory_type, response_memory_type_id,
+              response_byte_size, pinned_buffer + offset,
+              const_cast<void*>(response_buffer), stream_, &completion_queue));
+          async_task_ids.push_back(count);
+        } else {
+          bool cuda_used = false;
+          status = CopyBuffer(
+              response_output->Name(), pinned_memory_type, pinned_memory_id,
+              response_memory_type, response_memory_type_id, response_byte_size,
+              pinned_buffer + offset, const_cast<void*>(response_buffer),
+              stream_, &cuda_used);
+          need_sync_ |= cuda_used;
 
-        if (!status.IsOk()) {
+          if (!status.IsOk()) {
+            LOG_STATUS_ERROR(
+                InferenceResponse::SendWithStatus(
+                    std::move(*response), TRITONSERVER_RESPONSE_COMPLETE_FINAL,
+                    status),
+                "error sending TensorFlow response");
+          }
+        }
+      }
+
+      offset += response_byte_size;
+      count++;
+    }
+
+    size_t index = 0;
+    for (auto& pr : def.responses_) {
+      if (std::find(async_task_ids.begin(), async_task_ids.end(), index) !=
```

Comment:
Is this to determine if the response had an async copy sent. It is very confusingly named and implemented. Can you instead just send the "std::unique_ptr<InferenceResponse>* response" along with the task and include it in the completion queue. That will be a lot cleaner and simpler to understand.

Reply:
Yes. I could do that instead. I can get rid of async_task_ids entirely then.

Reply:
We can't generalize to add response in all cases the because there is a case a CopyBuffer call corresponds to reporting errors for multiple requests.

Reply:
This is still convoluted. It doesn't matter that you have to do something different elsewhere in the code. Here you should do the correct and clear thing and send the response along with the task and get it back in the completion queue. Also, it seems like you are assuming that the tasks complete in the same order as they are queued (and in the def.responses_ order). That is not necessarily true and will be fixed by passing the response.